src/file.cpp

```diff
@@ -946,21 +974,20 @@ namespace libtorrent
 	{
 #if defined TORRENT_WINDOWS && !defined TORRENT_MINGW
 #if TORRENT_USE_WSTRING
-		wchar_t cwd[TORRENT_MAX_PATH];
-		_wgetcwd(cwd, sizeof(cwd) / sizeof(wchar_t));
+		wchar_t* cwd { ::_wgetcwd(nullptr, 0) };
+		std::shared_ptr<wchar_t> holder(cwd, &std::free);
+		std::string ret { cwd == nullptr ? "" : convert_from_wstring(cwd) };
 #else
-		char cwd[TORRENT_MAX_PATH];
-		_getcwd(cwd, sizeof(cwd));
+		char* cwd { ::_getcwd(nullptr, 0) };
+		std::shared_ptr<char> holder(cwd, &std::free);
+		std::string ret { cwd == nullptr ? "" : convert_from_native(cwd) };
 #endif // TORRENT_USE_WSTRING
 #else
-		char cwd[TORRENT_MAX_PATH];
-		if (getcwd(cwd, sizeof(cwd)) == nullptr) return "/";
-#endif
-#if defined TORRENT_WINDOWS && !defined TORRENT_MINGW && TORRENT_USE_WSTRING
-		return convert_from_wstring(cwd);
-#else
-		return convert_from_native(cwd);
+		char* cwd { ::getcwd(nullptr, 0) };
+		std::shared_ptr<char> holder(cwd, &std::free);
+		std::string ret { cwd == nullptr ? "/" : convert_from_native(cwd) };
```

Comment:
I thinking that the last two lines here are nearly identical in all 3 cases, and may be a candidate for factoring out of the #ifdef check. The only difference is how to treat ``getcwd()`` returning null. Do you know under which conditions it does that? And is it really reasonable to silently ignore it, returning an empty string or "/"?

I realize this is the current behavior though. But would you mind throwing an ``lt::system_error()`` exception in case this fails? The error code (I believe) should be ``error_code(errno, generic_category());``